### ISSUE TITLE
Mean Field Dev - tempo.py

### ISSUE THREAD

**Issue by gefux** [tempoCollaboration/OQuPy#57]
Hi @piperfw !

As discussed, the `tempo.py` needs restructuring. See my suggestion below.

The logic is that we've pulled all common parts of `Tempo` and `TempoWithField` into `BaseTempo`. This leaves three things to do for each of them:
1. the preparation of the backend `._prepare_backend`,
2. the preparation of the propagators `._propagators()`, and 
3. the overall computation `.compute()`  

```python
class TempoParameters(BaseAPIClass):
    pass # Stays unchanged!

class BaseTempo(BaseAPIClass):
    def __init__(
            self,
            bath: Bath,
            parameters: TempoParameters,
            initial_state: ndarray,
            start_time: float,
            backend_config: Optional[Dict] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    def _influence(self, dk: int) -> ndarray:
        pass # ToDo

    def _time(self) -> float:
        pass # ToDo

    @property
    def dimension(self) -> ndarray:
        pass # ToDo

class Tempo(BaseTempo):
    def __init__(
            self,
            system: Union[System, TimeDependentSystem],
            bath: Bath,
            parameters: TempoParameters,
            initial_state: ndarray,
            start_time: float,
            backend_config: Optional[Dict] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    def _init_backend(self):
        pass # ToDo

    def _propagators(self, step: int) -> Tuple[ndarray, ndarray]:
        pass # ToDo

    def compute(
            self,
            end_time: float,
            progress_type: Text = None) -> Dynamics:
        pass # ToDo

class TempoWithField(BaseTempo):
    def __init__(
            self,
            system: TimeDependentSystemWithField,
            bath: Bath,
            parameters: TempoParameters,
            initial_state: ndarray,
            initial_field: complex,
            start_time: float,
            epsrel: Optional[float] = INTEGRATE_EPSREL,
            subdiv_limit: Optional[int] = SUBDIV_LIMIT,
            backend_config: Optional[Dict] = None,
            name: Optional[Text] = None,
            description: Optional[Text] = None) -> None:
        pass # ToDo

    def _init_backend(self):
        pass # ToDo

    def _propagators(
            self,
            step: int,
            state: ndarray,
            field: complex) -> Tuple[ndarray, ndarray]:
        pass # ToDo

    def compute(
            self,
            end_time: float,
            progress_type: Text = None) -> Tuple[DynamicsWithField]:
        pass # ToDo
```

**Comment by piperfw** [comment]
Hi @gefux,

I've fleshed out the proposed changes on [piperfw/TimeEvolvingMPO](https://github.com/piperfw/TimeEvolvingMPO/tree/mean-field-dev). 

Regarding the three methods you mention for each class:
1. `._prepare_backend()` has a lot of code in common between `Tempo` and `TempoWithField`. I don't currently see a neat way around this since they need to call different backend instance classes (with different function signatures). Let me know if you have any suggestions here.
2. `._propagators()` for each class are very different. Note #41 I should implement integration in `Tempo._propagators()` at some point in the future.
3. `.compute()` for each class look similar, but really are different due to the additional field being moved around in `TempoWithField`

Two other small comments:
- I removed reference to `backend: str` in the docstrings as this parameter is no longer used
- There is one additional helper function `_check_time` and method `_get_end_step_num_step` (currently in `BaseTempo`) to avoid a small amount of duplication between `.compute()` functions

When you have had a chance to review the changes we should discuss them and also a plan for `contractions.py` (other Todos: tests, docs).

**Comment by gefux** [comment]
Hi @piperfw,

I agree with all your comments. I think the almost-dublicate code in `._prepare_backend()` is quite bearable. One could make a method in `BaseTempo` that takes a `propagators` function and a `backend_class` to avoid any dublication, but I believe this harms the readability and simplicity more than it helps overall.

Some more comments:
- We'll need to change the input to `Tempo._system` to be of the type `Union[System, TimeDependentSystem]` and adopt `Tempo._propagators` to deal with the two different cases as a first step. 
- If possible please try to make pylint happy (line length, trailing whitespaces etc.) and give the appropriate adaptation of the tests a go.

Also, yes we should discuss `contractions.py` at some point. I think that this problem is luckily almost orthogonal to the rest, so we can deal with it independently.

**Comment by piperfw** [comment]
To resolve the issue with `Tempo._propagators` (without creating a new class `TimeDependentTempo`...) this method now checks what type of `self._system` it has. Let me know if you think that is reasonable for a permanent solution.
Since `Tempo` must be able to hand `TimeDependentSystem` as well, I've added a new test for this in `tempo_test.py`.

Overall, pylint is happy and all tests are passing _except_ those associated with calls to `contractions.py`, as is to be addressed. docs build too.

**Comment by gefux** [comment]
Yes, that's what I had in mind too. Thank you @piperfw!

**Comment by gefux** [comment]
changes are merged into the mean-field-dev branch #62
